src/ident/Ident.cc

```diff
@@ -243,9 +243,9 @@ Ident::Start(const Comm::ConnectionPointer &conn, IDCB * callback, void *data)
 
     conn->local.toUrl(key1, IDENT_KEY_SZ);
     conn->remote.toUrl(key2, IDENT_KEY_SZ);
-    const auto res = snprintf(key, sizeof(key), "%s,%s", key1, key2);
+    const int res = snprintf(key, sizeof(key), "%s,%s", key1, key2);
```

Comment:
IIRC it was a signed-unsigned comparison. Trying a build to double check

Reply:
It seems to be working. Maybe I was mistaken in thinking it wouldn't? reverting

Reply:
This is the error:

```
In file included from /home/kinkie/src/squid-wextra/src/base/CbcPointer.h:14,
                 from /home/kinkie/src/squid-wextra/src/base/AsyncJobCalls.h:13,
                 from /home/kinkie/src/squid-wextra/src/CommCalls.h:13,
                 from /home/kinkie/src/squid-wextra/src/comm.h:13,
                 from /home/kinkie/src/squid-wextra/src/ident/Ident.cc:14:
/home/kinkie/src/squid-wextra/src/ident/Ident.cc: In function ‘void Ident::Start(const ConnectionPointer&, void (*)(const char*, void*), void*)’:
/home/kinkie/src/squid-wextra/src/ident/Ident.cc:248:68: error: type qualifiers ignored on cast result type [-Werror=ignored-qualifiers]
  248 |     assert(static_cast<std::make_unsigned<decltype(res)>::type>(res) < sizeof(key));
      |                                                                    ^
/home/kinkie/src/squid-wextra/src/Debug.h:31:23: note: in definition of macro ‘assert’
   31 | #define assert(EX)  ((EX)?((void)0):xassert( # EX , __FILE__, __LINE__))
      |                       ^~
```

Reply:
We should figure out which "type qualifiers" are ignored. I wonder if it is `const`. Unfortunately, I cannot reproduce this with a couple of compilers (and I do not want to suggest inserting `const` in random places). What is your compiler name and version?

Reply:
I found a [compiler](https://cpp.godbolt.org/z/g3SGXj) that produces this warning -- GCC v10.1. AFAICT, we need to remove the `const` qualifier that decltype() preserves and the compiler warns us about (because the preserved `const` is not needed in this context, just like `const` for a function return type).

Here are a few solutions that compile in my tests:

```C++
typedef unsign_const<decltype(res)>::type unsigned_result;
assert(static_cast<unsigned_result>(res) < sizeof(key));
```

where we define `unsign_const` or, if you prefer, `make_a_const_type_unsigned` as follows (outside the function):

```C++
template <typename T>
using unsign_const = typename std::make_unsigned< typename std::remove_const<T>::type >;
```

We can make this prettier (and we should if this logic is reused!) using a helper function:

```C++
/// \returns the given non-negative parameter value after changing its type to unsigned
template <typename T>
constexpr typename unsign_const<T>::type
make_const_var_unsigned(T var)
{
    return static_cast< typename unsign_const<T>::type >(var);
}
```

With the above helper, our assert becomes very readable:

```C++
assert(make_const_var_unsigned(res) < sizeof(key));
```

You can see the above code in action (and test with different compilers) at https://cpp.godbolt.org/z/g3SGXj

Reply:
Disclaimer: It may be possible to simplify and/or generalize the above further. For example, it is possible that `const` disappears automatically from `T` when the helper function is called, simplifying the problem. I hope my sketches allow you to make progress, but I do not insist on this specific solution or claim that it is the final/best solution possible!

Reply:
Done. It looks nice and reuseable, so why not?

Reply:
Maybe easier to just remove the const from "const auto res" ?

Reply:
> Francesco: It looks nice and reuseable, so why not?

I agree.

> Amos: Maybe easier to just remove the const from "const auto res" ?

Yes, of course, it is easier. It is even easier to use `int` and `unsigned int` explicitly here. However, there is also significant value in following a good principle such as "use const and auto whenever possible". Every single violation is going to have negligible effect until the principle gets lost in the noise of violations, and good-meaning folks start adding them on purpose because the surrounding code "looks that way".

As you know, I did not insist on using these complex type manipulations, but I was glad to learn something new developing them, and I think they may be useful in other contexts some day (where they will be the only solution rather than just an annoying extra complexity), so I support adding/using them today. If they backfire, we can always make an exception and remove `const` or even `auto` in this use case, of course.

Reply:
@kinkie,  I am resolving this change request, but please move `typecast.h` to `src/base/TypeTraits.h`. We should avoid adding new code outside of `src/` (and should move existing "external" code into `src/` but that is another TODO).

Reply:
Done